### PR TITLE
JVM write permissions

### PR DESCRIPTION
**Pull request by brianhuffman**
This PR uses the new `crucible-jvm` write permissions added in GaloisInc/crucible#733 and GaloisInc/crucible#735 to fix the soundness issues #900 and #1066.

I'll keep this in a draft state until those crucible PRs are merged.

### CODE REVIEW

File: src/SAWScript/Interpreter.hs [hunk 1]
```diff
@@ -2695,6 +2695,54 @@ primitives = Map.fromList
...
+    [ "Declare that the indicated array (first argument) has an element"
+    , "(second argument) containing an unspecified value."
+    , ""
+    , "This lets users write partial specifications of JVM methods."
+    , "In the post-state section (after `jvm_execute_func`), it"
+    , "states that the method may modify the array element, but says"
+    , "nothing about the new value."
+    ]
+
+    , prim "jvm_modifies_array" "JVMValue -> JVMSetup ()"
```
Comment (RyanGlScott) [inline src/SAWScript/Interpreter.hs:2734]: Formatting nit: this `prim` line is indented by four spaces, while other `prim` lines in this list are indented by two spaces.

File: src/SAWScript/Interpreter.hs [hunk 2]
```diff
@@ -2695,6 +2695,54 @@ primitives = Map.fromList
 
     -- TODO: jvm_alloc_multiarray
 
+  , prim "jvm_modifies_field" "JVMValue -> String -> JVMSetup ()"
```
Comment (RyanGlScott) [inline src/SAWScript/Interpreter.hs:2698]: Are there test cases which use these new functions?

File: src/SAWScript/Crucible/JVM/Override.hs [hunk 3]
```diff
@@ -723,6 +723,10 @@ learnPointsTo opts sc cc spec prepost pt = do
          realTerm <- liftIO $ scVector sc ety_tm ts
          matchTerm sc cc loc prepost realTerm (ttTerm tt)
 
+    -- If the right-hand-side is 'Nothing', this is indicates a "modifies" declaration,
+    -- which should probably not appear in the pre-state section, and has no effect.
```
Comment (atomb) [inline src/SAWScript/Crucible/JVM/Override.hs:727]: Should we warn about this?

File: src/SAWScript/Crucible/JVM/Builtins.hs [hunk 4]
```diff
@@ -463,6 +487,8 @@ setupPrePointsTos mspec cc env pts mem0 = foldM doPointsTo mem0 pts
                  Just x -> pure x
              rhs' <- traverse (injectSetupVal . MS.SetupTerm) tts
              doEntireArrayStore sym mem lhs' rhs'
+        JVMPointsToArray _loc _lhs Nothing ->
+          pure mem -- This should probably never occur in the prestate section.
```
Comment (atomb) [inline src/SAWScript/Crucible/JVM/Builtins.hs:491]: Is it worth warning if this happens?

**Review by atomb — APPROVED** [review]

File: src/SAWScript/Crucible/JVM/Builtins.hs (hunk 4 above)
Comment (brianhuffman) [reply src/SAWScript/Crucible/JVM/Builtins.hs:491]: A `jvm_modifies` declaration in the pre-state section is basically meaningless. I could make that an error. But probably what I should do is make all the `jvm_modifies` declarations be recorded in the post-state section, no matter where they are written. I think that is closest to the behavior that a reasonable user would expect.

If I make that change, then I can make this particular line of code into a panic.

**Comment by brianhuffman** [comment]
It looks like there's a build failure caused by one of the crucible PRs that snuck in between my two crucible PRs. I'll have to make a separate bump-submodules PR to fix that, and merge it ahead of this one.

File: src/SAWScript/Interpreter.hs (hunk 2 above)
Comment (brianhuffman) [reply src/SAWScript/Interpreter.hs]: I'll add some test cases to demonstrate the new primitives.

Comment (brianhuffman) [reply src/SAWScript/Interpreter.hs]: The new test in `intTests/test_jvm_modifies` demonstrates all the new primitives.

File: src/SAWScript/Crucible/JVM/Builtins.hs (hunk 4 above)
Comment (brianhuffman) [reply src/SAWScript/Crucible/JVM/Builtins.hs]: I decided to just make `jvm_modifies` in the pre-state be an error. I extended `test_jvm_setup_errors` in the test suite to also test this new error condition, and that line of code is now a panic.

File: src/SAWScript/Crucible/JVM/Override.hs (hunk 3 above)
Comment (brianhuffman) [reply src/SAWScript/Crucible/JVM/Override.hs]: `jvm_modifies` in the pre-state section is now an error, so this can't happen.

**Pull request merged**